caluma/caluma_workflow/domain_logic.py

```diff
@@ -528,3 +528,66 @@ def post_resume(work_item, user, context=None):
         )
 
         return work_item
+
+
+class RedoWorkItemLogic:
+    @classmethod
+    def validate_for_redo(cls, work_item):
+        if work_item.status == models.WorkItem.STATUS_READY:
+            raise ValidationError("Ready work items can't be redone.")
+        ready_work_items = cls._find_ready_in_work_item_tree(
+            work_item, allowed_states=[models.WorkItem.STATUS_READY]
+        )
+        allowed = False
+        for wi in ready_work_items:
+            if work_item.task in wi.get_redoable():
+                allowed = True
+                break
+        if not allowed:
+            raise ValidationError("Workflow doesn't allow to redo this work item.")
+
+    @classmethod
+    def set_succeeding_work_item_status_redo(cls, work_item):
+        work_items = cls._find_ready_in_work_item_tree(work_item)
+        for wi in work_items:
+            wi.status = models.WorkItem.STATUS_REDO
+            wi.save()
+
+    @classmethod
+    def _find_ready_in_work_item_tree(cls, work_item, allowed_states=None, ready=None):
```

Comment:
If you'd convert this method into a generator, you'll improve two thing: The code gets simpler as you can use `yield from` instead of building up a list and passing it along. And also, it could improve performance, as we won't have to walk the tree further than needed once a workitem is found that allows the redo

Reply:
Nice, done.